src/lib/key-provider.c

```diff
@@ -24,55 +24,114 @@
  * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
 
+#include <assert.h>
+
 #include "key-provider.h"
 #include "pgp-key.h"
+#include "fingerprint.h"
+#include "types.h"
+#include "utils.h"
 #include <rekey/rnp_key_store.h>
 
 bool
-pgp_request_key(const pgp_key_provider_t *   provider,
-                const pgp_key_request_ctx_t *ctx,
-                pgp_key_t **                 key)
+rnp_key_matches_search(const pgp_key_t *key, const pgp_key_search_t *search)
 {
-    if (!provider || !provider->callback || !ctx || !key) {
+    if (!key) {
         return false;
     }
-    return provider->callback(ctx, key, provider->userdata);
+    switch (search->type) {
+    case PGP_KEY_SEARCH_KEYID:
+        return memcmp(key->keyid, search->by.keyid, PGP_KEY_ID_SIZE) == 0;
+    case PGP_KEY_SEARCH_FINGERPRINT:
+        return (key->fingerprint.length == search->by.fingerprint.length) &&
+               !memcmp(key->fingerprint.fingerprint,
+                       search->by.fingerprint.fingerprint,
+                       key->fingerprint.length);
+    case PGP_KEY_SEARCH_GRIP:
+        return !memcmp(key->grip, search->by.grip, PGP_FINGERPRINT_SIZE);
+    case PGP_KEY_SEARCH_USERID: {
+        for (unsigned i = 0; i < key->uidc; i++) {
+            if (!strcmp((char *) key->uids[i], search->by.userid)) {
```

Comment:
Maybe use strstr/strcasestr here? Or do we need exact match here?

Reply:
My feeling is it's best to go for an exact match within the key provider itself. An application using the FFI can always use the id iterator to do fancier matces like case-insensitive/regex userid, or partial keyid, and so on. (Or we could implement an FFI search function with a callback to check custom criteria, etc.)

Reply:
@ni4 if it’s OK feel free to merge it. Thanks!